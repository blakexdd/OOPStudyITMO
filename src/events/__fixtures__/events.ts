import { CategoriesEnum } from '@src/events/interfaces/categories.interface';

export const firstEvent = {
  name: 'fistEvent',
  category: CategoriesEnum.OFFLINE_EVENT,
  price: 100,
  purchaseLink: 'https://localhost',
  description: 'Some event',
  link: 'https://localhost',
  age: 18,
  maxPerson: 45,
};
export const secondEvent = {
  name: 'second',
  category: CategoriesEnum.OFFLINE_EVENT,
  price: 2,
  purchaseLink: 'https://123',
  description: 'Second',
  link: 'https://234',
  age: 0,
  maxPerson: 0,
};
