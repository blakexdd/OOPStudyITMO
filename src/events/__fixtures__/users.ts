export const firstUser = {
  name: 'first',
  password: 'F^U3Lk0QU5',
  email: 'first@mail.ru',
};
export const secondUser = {
  name: 'seconds',
  password: 'F^U3Lk0QU5',
  email: 'second@mail.ru',
};
