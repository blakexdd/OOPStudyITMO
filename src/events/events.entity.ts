import { CategoriesEnum } from '@src/events/interfaces/categories.interface';
import { UserEntity } from '@src/users/user.entity';
import { ApiProperty } from '@nestjs/swagger';

export class EventEntity {
  @ApiProperty({ example: 'Example event' })
  name: string;

  @ApiProperty({ example: CategoriesEnum.OFFLINE_EVENT })
  category: CategoriesEnum;

  @ApiProperty({ example: 100 })
  price: number;

  @ApiProperty({ example: 'https://event.com' })
  purchaseLink: string;

  @ApiProperty({ example: 'Some event description' })
  description: string;

  @ApiProperty({ example: 'https://event.com' })
  link: string;

  @ApiProperty({ example: 18 })
  age: number;

  @ApiProperty({ example: 5 })
  maxPerson: number;

  @ApiProperty()
  owner: UserEntity;

  @ApiProperty({ type: [UserEntity] })
  participants: UserEntity[];

  constructor(partial: Partial<EventEntity>) {
    this.name = partial.name;
    this.category = partial.category;
    this.price = partial.price;
    this.purchaseLink = partial.purchaseLink;
    this.description = partial.description;
    this.link = partial.link;
    this.age = partial.age;
    this.maxPerson = partial.maxPerson;
    this.owner = new UserEntity(partial.owner);
    this.participants = partial.participants.map((user) => {
      return new UserEntity(user);
    });
  }
}
