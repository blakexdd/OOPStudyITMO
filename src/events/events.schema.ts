import { SchemaObject } from 'neode';
import { CategoriesEnum } from '@src/events/interfaces/categories.interface';

const EventsSchema: SchemaObject = {
  id: {
    type: 'uuid',
    primary: true,
    required: true,
  },
  name: {
    type: 'string',
    required: true,
    unique: true,
  },
  category: {
    type: 'string',
    required: true,
    default: CategoriesEnum.OFFLINE_EVENT,
  },
  price: {
    type: 'number',
    default: 0,
  },
  purchaseLink: {
    type: 'string',
  },
  description: {
    type: 'string',
  },
  link: {
    type: 'string',
  },
  age: {
    type: 'number',
  },
  maxPerson: {
    type: 'number',
  },
  owner: {
    type: 'relationship',
    target: 'User',
    relationship: 'OWNER',
    direction: 'in',
    properties: {
      id: 'string',
    },
    eager: true,
    cascade: 'delete',
  },
  lineups: {
    type: 'relationships',
    target: 'Lineup',
    relationship: 'LINEUP',
    direction: 'out',
    properties: {
      id: 'string',
    },
    eager: true,
  },
};

export default EventsSchema;
