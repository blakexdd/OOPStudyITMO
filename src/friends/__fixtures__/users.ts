export const firstUser = {
  name: 'first',
  password: 'F^U3Lk0QU5',
  email: 'first@mail.ru',
};
export const secondUser = {
  name: 'seconds',
  password: 'F^U3Lk0QU5',
  email: 'second@mail.ru',
};
export const thirdUser = {
  name: 'third',
  password: 'F^U3Lk0QU5',
  email: 'third@mail.ru',
};
