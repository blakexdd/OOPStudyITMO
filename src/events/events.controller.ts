import {
  Controller,
  Post,
  Get,
  Put,
  Delete,
  Body,
  Param,
  UseGuards,
  Req,
  UseInterceptors,
  ClassSerializerInterceptor,
} from '@nestjs/common';
import { JwtAuthGuard } from '@src/auth/strategy/jwt-auth-guard';
import { CreateEventDto } from '@src/events/dto/create-event.dto';
import { UpdateEventDto } from '@src/events/dto/update-event.dto';
import { AuthenticatedUserRequest } from '@src/generic.interface';
import { EventsService } from '@src/events/events.service';
import { ApiBearerAuth, ApiTags, ApiCreatedResponse } from '@nestjs/swagger';
import { EventEntity } from '@src/events/events.entity';
import { CreatedDto } from '@src/generic.dto';

@ApiBearerAuth()
@ApiTags('Events')
@Controller('events')
export class EventsController {
  constructor(private eventsService: EventsService) {}

  @UseGuards(JwtAuthGuard)
  @Post()
  @ApiCreatedResponse({
    description: 'Record successfully created',
    type: CreatedDto,
  })
  async create(
    @Req() req: AuthenticatedUserRequest,
    @Body() createEventDto: CreateEventDto,
  ) {
    return this.eventsService.create({
      event: createEventDto,
      ownerId: req.user.id,
    });
  }

  @UseGuards(JwtAuthGuard)
  @UseInterceptors(ClassSerializerInterceptor)
  @Get(':id')
  @ApiCreatedResponse({
    description: 'Record successfully retrieved',
    type: EventEntity,
  })
  async findOne(@Param('id') id: string): Promise<EventEntity> {
    const eventDb = await this.eventsService.findOne({ eventId: id });
    return new EventEntity({
      ...eventDb,
      //@ts-ignore
      owner: eventDb.owner.node,
    });
  }

  @UseGuards(JwtAuthGuard)
  @UseInterceptors(ClassSerializerInterceptor)
  @Put(':id')
  @ApiCreatedResponse({
    description: 'Record successfully updated',
    type: EventEntity,
  })
  async update(
    @Param('id') id: string,
    @Body() updateEventDto: UpdateEventDto,
    @Req() req: AuthenticatedUserRequest,
  ): Promise<EventEntity> {
    const eventDb = await this.eventsService.update({
      event: updateEventDto,
      eventId: id,
      user: req.user,
    });

    return new EventEntity({
      ...eventDb,
      //@ts-ignore
      owner: eventDb.owner.node,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Delete(':id')
  async remove(@Param('id') id: string, @Req() req: AuthenticatedUserRequest) {
    await this.eventsService.remove({ eventId: id, user: req.user });
  }
}
