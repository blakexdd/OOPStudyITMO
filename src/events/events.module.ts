import { Module } from '@nestjs/common';
import { EventsController } from '@src/events/events.controller';
import { EventsService } from '@src/events/events.service';
import { CaslModule } from '@src/casl/casl.module';
import { NeodeModule } from '@src/neo4j/neo4j.module';
import EventsSchema from '@src/events/events.schema';
import UserSchema from '@src/users/users.schema';

@Module({
  imports: [
    CaslModule,
    NeodeModule.forFeature({ Event: EventsSchema, User: UserSchema }),
  ],
  controllers: [EventsController],
  providers: [EventsService],
})
export class EventsModule {}
