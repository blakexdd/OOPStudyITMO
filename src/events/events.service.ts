import { HttpException, Inject, Injectable } from '@nestjs/common';
import { Action } from '@src/generic.interface';
import { CaslEventsAbilityFactory } from '@src/casl/casl-events-ability.factory';
import Neode from 'neode';
import { FullEventInterface } from '@src/events/interfaces/events.interface';
import { CreateEventInterface } from '@src/events/interfaces/create-event.interface';
import { UpdateEventInterface } from '@src/events/interfaces/update-event.interface';
import { RemoveEventInterface } from '@src/events/interfaces/remove-event.interface';
import { FindOneInterface } from '@src/events/interfaces/find-one.interface';
import { uuid } from 'uuidv4';

@Injectable()
export class EventsService {
  constructor(
    @Inject('Connection') private readonly neode: Neode,
    private caslEventsAbilityFactory: CaslEventsAbilityFactory,
  ) {}

  async create(params: CreateEventInterface) {
    const id: string = uuid();
    const createEventQuery = `
      MATCH (u: User {id: $userId})
      CREATE (e: Event $event)
      CREATE (u)-[rel:OWNER]->(e)
    `;
    await this.neode.cypher(createEventQuery, {
      userId: params.ownerId,
      event: { ...params.event, id },
    });

    return { id };
  }

  async update(params: UpdateEventInterface) {
    const ownerId = await this.getOwnerId(params.eventId);

    await this.caslEventsAbilityFactory.hasAbility(
      params.user,
      Action.UPDATE,
      ownerId,
    );

    const event = await this.neode.merge<FullEventInterface>(
      'Event',
      params.event,
    );

    return event.toJson();
  }

  async remove(params: RemoveEventInterface) {
    const ownerId = await this.getOwnerId(params.eventId);

    await this.caslEventsAbilityFactory.hasAbility(
      params.user,
      Action.DELETE,
      ownerId,
    );

    const query = this.neode.query();
    const eventModel = this.neode.model('Event');

    await query
      .match('event', eventModel)
      .where('event.id', params.eventId)
      .detachDelete('event')
      .execute();
  }

  async findOne(params: FindOneInterface) {
    const event = await this.neode.first<FullEventInterface>(
      'Event',
      'id',
      params.eventId,
    );

    if (!event) {
      throw new HttpException('Event not found', 404);
    }

    return event.toJson();
  }

  private async getOwnerId(eventId: string) {
    const query = this.neode.query();
    const eventModel = this.neode.model('Event');
    const userModel = this.neode.model('User');

    const res = await query
      .match('event', eventModel)
      .where('event.id', eventId)
      .relationship('OWNER', 'in', 'rel', 1)
      .to('owner', userModel)
      .return('owner')
      .execute();

    if (res.records.length) {
      return res.records[0].get('owner').properties.id;
    }

    return undefined;
  }
}
