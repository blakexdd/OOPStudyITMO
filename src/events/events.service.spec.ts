import { Test, TestingModule } from '@nestjs/testing';
import { EventsService } from './events.service';
import { ConfigModule } from '@nestjs/config';
import { NeodeModule } from '@src/neo4j/neo4j.module';
import UserSchema from '@src/users/users.schema';
import { UsersService } from '@src/users/users.sevice';
import Neode from 'neode';
import { CaslModule } from '@src/casl/casl.module';
import EventsSchema from '@src/events/events.schema';
import { omit } from 'lodash';
import { HttpException } from '@nestjs/common';
import { Action } from '@src/generic.interface';
import { firstEvent, secondEvent } from '@src/events/__fixtures__/events';
import { firstUser, secondUser } from '@src/events/__fixtures__/users';

describe('EventsService', () => {
  let eventsService: EventsService;
  let usersService: UsersService;
  let neode: Neode;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
        }),
        NeodeModule.forFeature({ Event: EventsSchema, User: UserSchema }),
        CaslModule,
      ],
      providers: [EventsService, UsersService],
    }).compile();

    eventsService = module.get<EventsService>(EventsService);
    usersService = module.get<UsersService>(UsersService);
    neode = module.get<Neode>('Connection');
    await Promise.all([neode.deleteAll('User'), neode.deleteAll('Event')]);
  });

  it('should be defined', () => {
    expect(eventsService).toBeDefined();
  });

  it('create event', async () => {
    const firstUserDb = await usersService.save(firstUser);

    const { id } = await eventsService.create({
      event: firstEvent,
      ownerId: firstUserDb.id,
    });
    const eventDb = await eventsService.findOne({ eventId: id });
    expect(omit(eventDb, ['id', '_id', 'owner'])).toMatchSnapshot();
  });

  it('update event', async () => {
    const firstUserDb = await usersService.save(firstUser);

    const { id } = await eventsService.create({
      event: firstEvent,
      ownerId: firstUserDb.id,
    });

    const eventDb = await eventsService.update({
      eventId: id,
      event: secondEvent,
      user: firstUserDb,
    });
    expect(omit(eventDb, ['id', '_id', 'owner'])).toMatchSnapshot();
  });

  test('update event of another user', async () => {
    const firstUserDb = await usersService.save(firstUser);
    const secondUserDb = await usersService.save(secondUser);

    const { id } = await eventsService.create({
      event: firstEvent,
      ownerId: firstUserDb.id,
    });

    await expect(
      eventsService.update({
        eventId: id,
        event: secondEvent,
        user: secondUserDb,
      }),
    ).rejects.toThrow(new HttpException(`Can not ${Action.UPDATE} event`, 403));
  });

  it('delete event', async () => {
    const firstUserDb = await usersService.save(firstUser);

    const { id } = await eventsService.create({
      event: firstEvent,
      ownerId: firstUserDb.id,
    });

    await eventsService.remove({ eventId: id, user: firstUserDb });
    await expect(eventsService.findOne({ eventId: id })).rejects.toThrow(
      new HttpException('Event not found', 404),
    );
  });

  test('delete event of another user', async () => {
    const firstUserDb = await usersService.save(firstUser);
    const secondUserDb = await usersService.save(secondUser);

    const { id } = await eventsService.create({
      event: firstEvent,
      ownerId: firstUserDb.id,
    });

    await expect(
      eventsService.remove({
        eventId: id,
        user: secondUserDb,
      }),
    ).rejects.toThrow(new HttpException(`Can not ${Action.DELETE} event`, 403));
  });
});
