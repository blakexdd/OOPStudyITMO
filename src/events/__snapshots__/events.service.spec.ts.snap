// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsService Update single event field 1`] = `
Object {
  "_labels": Array [
    "Event",
  ],
  "age": 18,
  "category": "OFFLINE_EVENT",
  "description": "Some event",
  "lineups": Array [],
  "link": "https://localhost",
  "maxPerson": 45,
  "name": "Changed name",
  "price": 0,
  "purchaseLink": "https://localhost",
}
`;

exports[`EventsService create event 1`] = `
Object {
  "_labels": Array [
    "Event",
  ],
  "age": 18,
  "category": "OFFLINE_EVENT",
  "description": "Some event",
  "lineups": Array [],
  "link": "https://localhost",
  "maxPerson": 45,
  "name": "fistEvent",
  "price": 100,
  "purchaseLink": "https://localhost",
}
`;

exports[`EventsService update event 1`] = `
Object {
  "_labels": Array [
    "Event",
  ],
  "age": 0,
  "category": "OFFLINE_EVENT",
  "description": "Second",
  "lineups": Array [],
  "link": "https://234",
  "maxPerson": 0,
  "name": "second",
  "price": 2,
  "purchaseLink": "https://123",
}
`;
